README.md
# selenium_tkit

> This package is under development, it should be used with that in mind, not every system may be compatible

[![PyPI version][pypi-image]][pypi-url]
[![GitHub stars][stars-image]][stars-url]
[![Support Python versions][versions-image]][versions-url]



## Getting started

You can [get `selenium_tkit` from PyPI](https://pypi.org/project/selenium_tkit),
which means it's easily installable with `pip`:

```bash
python -m pip install selenium_tkit
```


## Example usage

```python
```



## Changelog

Refer to the [CHANGELOG.md](https://github.com/henriquelino/selenium_tkit/blob/main/CHANGELOG.md) file.



<!-- Badges -->

[pypi-image]: https://img.shields.io/pypi/v/selenium_tkit
[pypi-url]: https://pypi.org/project/selenium_tkit/


[stars-image]: https://img.shields.io/github/stars/henriquelino/selenium_tkit
[stars-url]: https://github.com/henriquelino/selenium_tkit

[versions-image]: https://img.shields.io/pypi/pyversions/selenium_tkit
[versions-url]: https://pypi.org/project/selenium_tkit/



CHANGELOG.md

# v0.1.0.3 -- 2023-08-03

## Changed

- Added webdriver_manager 4.0.0 support for chrome

# v0.1.0.2 -- 2023-02-16

## Changed

- Log levels for a bunch of functions was critical, now changed to debug

- ReusableChrome now accepts kwargs

## Fixed

- wait_find_elements wasn't working because variable names were wrong

# v0.1.0.1 -- 2023-01-08

## Added

- github publish action

# v0.1.0 -- 2023-01-08

## Added

- Package published


tests/test_chrome.py
import json
import logging
import os
import sys
import time
from pathlib import Path
import shutil

import pytest
from src.selenium_tkit.chrome import CreateChrome, ReusableChrome

BASE_DIR = Path(__file__).parent.resolve()
WEBDRIVER_PATH = BASE_DIR / "webdriver"
ID_PATH = BASE_DIR / "id.json"

def teardown():
    ReusableChrome.end_all_driver_processes()
    if ID_PATH.exists():
        ID_PATH.unlink()
        
    if WEBDRIVER_PATH.exists():
        shutil.rmtree(str(WEBDRIVER_PATH))
        
    

logging.getLogger("selenium").setLevel(logging.WARNING)
logging.getLogger("urllib3").setLevel(logging.WARNING)

if sys.platform == "win32":
    # without this tox couldn't find chrome binary
    os.environ.update({"PROGRAMW6432": "C:\\Program Files",})


class TestDisposable:
    
    def test_basic_create(self):
        teardown()
        chrome = CreateChrome(WEBDRIVER_PATH)
        chrome.begin()
        chrome.open_url("https://www.google.com/")
        assert chrome.current_url == "https://www.google.com/"
        chrome.quit()


class TestReusable:
    
    def test_create_wo_patch(self):
        teardown()
        chrome = ReusableChrome(WEBDRIVER_PATH, ID_PATH, apply_patch=False)
        ret = chrome.begin()
        assert ret is True
        assert chrome.open_url("https://www.google.com/") is True
        chrome.quit()
        assert chrome.end_all_driver_processes() is True
        
    def test_end_all_chrome_processes_wo_chrome(self):
        assert ReusableChrome.end_all_driver_processes() is True
        
    def test_end_all_chrome_processes_w_chrome(self):
        chrome = ReusableChrome(str(WEBDRIVER_PATH), ID_PATH)
        ret = chrome.begin()
        assert ret is True
        assert chrome.end_all_driver_processes() is True
        
    def test_create(self):
        teardown()
        chrome = ReusableChrome(WEBDRIVER_PATH, ID_PATH)
        ret = chrome.begin()
        assert ret is True
        assert chrome.open_url("https://www.google.com/") is True
        chrome.quit()
        assert chrome.end_all_driver_processes() is True
        
    def test_reuse(self):
        """uses the same session from the last test, with same id.json, this does not run teardown"""
        chrome = ReusableChrome(WEBDRIVER_PATH, ID_PATH)
        ret = chrome.begin()
        assert ret is True
        assert chrome.open_url("https://www.google.com/") is True
        chrome.quit()
        assert chrome.end_all_driver_processes() is True
        
        
    def test_broken_id_json(self):
        """uses the same session from the last test, with same id.json, this does not run teardown"""
        with ID_PATH.open('w') as f:
            f.write("")
            
        chrome = ReusableChrome(WEBDRIVER_PATH, ID_PATH)
        ret = chrome.begin()
        assert ret is True
        
    

src/selenium_tkit/edge.py
import logging
import re
from pathlib import Path
from typing import Optional, Union

from selenium.webdriver.edge.options import Options as EdgeOptions
from selenium.webdriver.edge.service import Service as EdgeService
from webdriver_manager.microsoft import EdgeChromiumDriverManager

from .custom_webdriver import CustomWebDriver

logger = logging.getLogger(__name__)


class CreateEdge(CustomWebDriver):

    def __init__(
        self,
        driver_path: Union[Path, str],  # yapf: disable
        implicity_wait: int = 0,
        port: int = 64900,
        options: Optional[EdgeOptions] = None,
        **kwargs,
    ) -> None:

        if isinstance(driver_path, Path):
            driver_path = str(driver_path)
        self.driver_path = EdgeChromiumDriverManager(path=driver_path).install()

        self.implicity_wait = implicity_wait
        self.port = port
        self.options = options
        return

    def begin(self):
        # TODO: quando o data-dir já está em execução, dá erro, precisa corrigr
        # ? testar se o data-dir está em uso, se sim, criar um novo?
        # ? ou qual ação tomar?

        serv = EdgeService(executable_path=self.driver_path, port=self.port)

        super().__init__(service=serv, options=self.options)

        self.implicitly_wait(self.implicity_wait)

        return True


def create_edge():
    """Este é apenas uma função exemplo!"""
    # import PythonUtils.log as log

    # log.create_logger(level=10)
    logging.getLogger("selenium").setLevel(logging.WARNING)
    logging.getLogger("urllib3").setLevel(logging.WARNING)

    # CustomChrome.end_all_chrome_processes()
    edge_configs = {
        "driver_path": "{APP_PATH}\chromedriver",
        "id_path": "{APP_PATH}\chromedriver\chrome_id.json",
        "port": 65000,
        "implicity_wait": 0,
        "attach_retries": 2,
        "new_console": True,
    }

    edge_options = {}

    # --------------------------------------------------

    options = EdgeOptions()

    if edge_options:
        for arg in edge_options["arguments"]:
            options.add_argument(arg)

        # --------------------------------------------------

        for k, v in edge_options["preferences"].items():
            if isinstance(v, str):
                v = v.replace("\n", "")
                v = re.sub(" +", " ", v)
            options.add_experimental_option(k, v)

    logger.critical(f"{options.arguments = }")
    logger.critical(f"{options.capabilities = }")

    # --------------------------------------------------

    driver = CreateEdge(**edge_configs, options=options)  # ok
    if not driver.begin():
        logger.critical(f"Falha ao criar edge driver")
        return False

    # --------------------------------------------------

    driver.rotate_user_agent()

    driver.set_navigator_to_undefined()

    # driver.refresh()

    # corrige a exception
    # unknown command: session/51ac9879f26bdb921197203660769456/se/file
    driver._is_remote = False

    return driver


src/selenium_tkit/firefox.py
import json
import logging
import os
import re
from pathlib import Path
from typing import Optional, Union

import psutil
from selenium.webdriver import Firefox
from selenium.webdriver.firefox.options import Options as FirefoxOptions
from selenium.webdriver.firefox.service import Service as FirefoxService
from selenium.webdriver.remote.webdriver import WebDriver
from webdriver_manager.firefox import GeckoDriverManager

from .custom_webdriver import CustomWebDriver

logger = logging.getLogger(__name__)


class ReusableFirefox(CustomWebDriver):

    def __init__(
        self,
        driver_path: Union[Path,
                           str],
        id_path: Union[Path,
                       str],
        implicity_wait: int = 0,
        port: int = 65000,
        options: Optional[FirefoxOptions] = None,
        attach_retries: int = 2,
        new_console: bool = True,
    ):

        if isinstance(driver_path, Path):
            driver_path = str(driver_path)
        self.driver_path = GeckoDriverManager(path=driver_path).install()

        self.id_path = id_path
        self.implicity_wait = implicity_wait
        self.port = port
        self.options = options
        self.attach_retries = attach_retries
        self.new_console = new_console
        return

    @classmethod
    def end_all_driver_processes(self):
        for p in psutil.process_iter():
            if "firefox" in p.name():
                logger.critical(f"Encerrando processo: '{p.name()}'")
                os.system(f"taskkill /f /t /im {p.name()}")

        for p in psutil.process_iter():
            if "firefox" in p.name():
                return False
        else:
            return True

    def begin(self) -> Union[WebDriver, bool]:
        """Realiza o fluxo de criação de um driver usável
        1. Cria os arquivos necessários (chromedriver.exe e id.json)
        2. Tenta utilizar um chrome já aberto
        2.1. Caso não consiga, encerra os processos do chrome abertos
        2.2. Re-abre os chromes, caso seja necessário, atualiza a versão
        3. Após obter um driver utilizável, salva as informações (command executor e ID) para ser reutilizado

        Returns
        ------
        `driver` : WebDriver
            instância do chromedriver
        `False` : bool
            Falha ao criar o chrome
        """

        # se o ID não existir...
        if not os.path.exists(self.id_path):
            logger.critical(f"Arquivo {self.id_path} não existe, criando um novo...")
            with open(self.id_path, "w", encoding="utf-8") as json_file:
                # cria um arquivo json com um dict vazio
                json.dump(dict(), json_file, indent=4)
            logger.critical(f"Arquivo '{self.id_path}' foi criado")

        # -----------------------------------
        # carrega o command executor e o session ID do arquivo id.json
        with open(self.id_path, "r") as json_file:
            try:
                chrome_ids = json.load(json_file)
            except json.decoder.JSONDecodeError:
                chrome_ids = dict()

        logger.debug(f"Arquivo de ID trouxe: '{chrome_ids}'")

        self.last_session_id = chrome_ids.get("session_id", None)
        self.last_command_executor = chrome_ids.get("command_executor", None)

        # -----------------------------------

        for tentativa_atual in range(self.attach_retries):
            logger.critical(f"Tentando criar driver {tentativa_atual+1}/{self.attach_retries}")
            attached = False

            # se tiver command_executor, session_id e um chrome existir, tenta usar uma sessão já aberta
            if (self.last_command_executor and self.last_session_id and self.any_chrome_process_exists):
                attached = self._attach()

            # se não tiver um dos três pra dar attach, então precisa reabrir o chrome
            if not attached:
                self.end_all_driver_processes()  # encerra os chromes em execução

            if attached:
                # se conseguiu dar attach em qualquer momento, sai do loop, se não continua
                break
        else:
            logger.critical(f"Não consegui criar um driver após '{self.attach_retries}' tentativas")
            return False

        # -----------------------------------
        # salva as informações da nova sessão do chromedriver
        chrome_configs = {
            "command_executor": self.command_executor._url,
            "session_id": self.session_id,
        }

        with open(self.id_path, "w", encoding="utf-8") as json_file:
            json.dump(chrome_configs, json_file, indent=4)

        logger.debug(f"Arquivo ID foi atualizado com: {chrome_configs}")

        return True


class CreateFirefox(CustomWebDriver, Firefox):

    def __init__(
        self,
        driver_path: Union[Path,
                           str],
        implicity_wait: int = 0,
        port: int = 64900,
        options: Optional[FirefoxOptions] = None,
        **kwargs,
    ) -> None:

        if isinstance(driver_path, Path):
            driver_path = str(driver_path)
        self.driver_path = GeckoDriverManager(path=driver_path).install()

        self.implicity_wait = implicity_wait
        self.port = port
        self.options = options
        return

    def begin(self):
        # TODO: quando o data-dir já está em execução, dá erro, precisa corrigr
        # ? testar se o data-dir está em uso, se sim, criar um novo?
        # ? ou qual ação tomar?

        #  options=self.options

        Firefox.__init__(self, service=FirefoxService(self.driver_path), options=self.options)

        # super().__init__()
        self.implicitly_wait(self.implicity_wait)

        return True


def create_firefox():
    """Este é apenas uma função exemplo!"""
    # import log

    # log.create_logger(level=10)
    logging.getLogger("selenium").setLevel(logging.WARNING)
    logging.getLogger("urllib3").setLevel(logging.WARNING)

    # CustomChrome.end_all_chrome_processes()
    firefox_configs = {
        "driver_path": "{APP_PATH}\chromedriver",
        "id_path": "{APP_PATH}\chromedriver\chrome_id.json",
        "port": 65000,
        "implicity_wait": 0,
        "attach_retries": 2,
        "new_console": True,
    }

    firefox_options = {
        "arguments": [
            # "--headless"
        ],
        # http://kb.mozillazine.org/Category:Preferences
        "preferences": {
            "browser.helperApps.neverAsk.saveToDisk": """application/csv,
                                                       application/excel,
                                                       application/vnd.ms-excel,
                                                       application/vnd.msexcel,,
                                                       application/xml,
                                                       application/octet-stream,
                                                       text/anytext,
                                                       text/comma-separated-values,
                                                       text/csv,
                                                       text/plain,
                                                       text/x-csv,,
                                                       text/xml
                                                       application/x-csv,
                                                       text/x-comma-separated-values,
                                                       text/tab-separated-values,
                                                       image/jpeg,
                                                       data:text/csv""",
            "browser.helperApps.neverAsk.openFile": """application/csv,
                                                     application/excel,
                                                     application/vnd.ms-excel,
                                                     application/vnd.msexcel,
                                                     text/anytext,
                                                     text/comma-separated-values,
                                                     text/csv,
                                                     text/plain,
                                                     text/x-csv,
                                                     application/x-csv,
                                                     text/x-comma-separated-values,
                                                     text/tab-separated-values,
                                                     data:text/csv,
                                                     application/xml,
                                                     text/plain,
                                                     text/xml,
                                                     image/jpeg,
                                                     application/octet-stream,
                                                     data:text/csv""",
            "browser.download.manager.showWhenStarting": False,
            "browser.helperApps.alwaysAsk.force": False,
            "browser.download.useDownloadDir": True,
            "dom.file.createInChild": True,
        },
        "extensions": [  # uma pasta com arquivos ou cada arquivo separado
            # r"D:\git\ITGreen\pythonutils\PythonUtils\chromedriver\extensions"
            # r"D:\git\ITGreen\pythonutils\PythonUtils\chromedriver\extensions\CRX-Extractor-Downloader.crx",
            # r"D:\git\ITGreen\pythonutils\PythonUtils\chromedriver\extensions\uBlock-Origin.crx",
        ],
    }

    # --------------------------------------------------

    options = FirefoxOptions()

    for arg in firefox_options["arguments"]:
        options.add_argument(arg)

    # --------------------------------------------------

    for k, v in firefox_options["preferences"].items():
        if isinstance(v, str):
            v = v.replace("\n", "")
            v = re.sub(" +", " ", v)
        options.set_preference(k, v)

    # --------------------------------------------------

    driver = CreateFirefox(**firefox_configs, options=options)  # ok
    if not driver.begin():
        logger.critical(f"Falha ao criar chromedriver")
        return False

    # --------------------------------------------------

    driver.rotate_user_agent()

    driver.set_navigator_to_undefined()

    # driver.refresh()

    # corrige a exception
    # unknown command: session/51ac9879f26bdb921197203660769456/se/file
    driver._is_remote = False

    return driver


src/selenium_tkit/chrome.py
import json
import logging
import os
import re
from pathlib import Path
from subprocess import CREATE_NEW_CONSOLE, CREATE_NO_WINDOW, Popen
from typing import Optional, Union

import psutil
from selenium.common.exceptions import (InvalidSessionIdException,
                                        SessionNotCreatedException,
                                        WebDriverException)
from selenium.webdriver import Chrome
from selenium.webdriver.chrome.options import Options as ChromeOptions
from selenium.webdriver.chrome.service import Service as ChromeService
from selenium.webdriver.remote.webdriver import WebDriver
from undetected_chromedriver.patcher import Patcher
from urllib3.exceptions import MaxRetryError
from webdriver_manager.chrome import ChromeDriverManager
from webdriver_manager.core.driver_cache import DriverCacheManager

from .custom_webdriver import CustomWebDriver

# --------------------------------------------------

logger = logging.getLogger(__name__)

# --------------------------------------------------

class ReusableChrome(CustomWebDriver):

    def __init__(
        self,
        driver_path: Union[Path, str],
        id_path: Union[Path, str],
        implicity_wait: int = 0,
        port: int = 65000,
        options: Optional[ChromeOptions] = None,
        attach_retries: int = 2,
        new_console: bool = True,
        apply_patch: bool = True,
        **kwargs
    ) -> None:  # yapf: disable
        """Cria um chromedriver que pode ser reutilizado após encerramento do Python

        ---
        Parameters
        ------
        `driver_path` : Path, str
            Caminho até o .exe do chromedriver
        `id_path` : Path, str
            Caminho até o arquivo .json que será salvo o id + command_executor
        `implicity_wait` : int
            Quando executar os scripts, esperar implicitamente por N segundos
        `port` : int
            Porta em que o chromedriver será aberta
        `options` : Options, None
            Opcional, parâmetros adicionais do webdriver
        `attach_retries` : int
            Quantas vezes tentar criar o driver
        `new_console` : bool
            ao criar novo chromedriver, abrir em um novo console ou não
        ---
        ### Exemplos:

        ```
        from PythonUtils.utils import BASE_DIR

        chrome_configs={
            "driver_path": f"{BASE_DIR}\\chromedriver\\chromedriver.exe",
            "id_path": f"{BASE_DIR}\\chromedriver\\chrome_id.json"
        }

        driver = CustomChrome(**chrome_configs)
        if not driver.begin():
            logger.critical(f"Falha ao criar chromedriver")
            exit()

        driver.open_url("https://www.google.com")
        ```
        """

        if isinstance(driver_path, Path):
            driver_path = str(driver_path)
            cache = DriverCacheManager(driver_path)
        self.driver_path = ChromeDriverManager(cache_manager=cache).install()

        if apply_patch:
            patcher = Patcher(executable_path=self.driver_path, force=False)
            patcher.auto()

            self.driver_path = patcher.executable_path

        self.id_path = id_path
        self.implicity_wait = implicity_wait
        self.port = port
        self.options = options
        self.attach_retries = attach_retries
        self.new_console = new_console

        return

    @property
    def any_chrome_process_exists(self):
        for p in psutil.process_iter():
            if "chromedriver.exe" in p.name():
                return True
        else:
            return False

    @classmethod
    def end_all_driver_processes(self):
        for p in psutil.process_iter():
            if "chrome" in p.name():
                logger.critical(f"Encerrando processo: '{p.name()}'")
                os.system(f"taskkill /f /t /im {p.name()}")

        for p in psutil.process_iter():
            if "chrome" in p.name():
                return False
        else:
            return True

    def begin(self) -> Union[WebDriver, bool]:
        """Realiza o fluxo de criação de um driver usável
        1. Cria os arquivos necessários (chromedriver.exe e id.json)
        2. Tenta utilizar um chrome já aberto
        2.1. Caso não consiga, encerra os processos do chrome abertos
        2.2. Re-abre os chromes, caso seja necessário, atualiza a versão
        3. Após obter um driver utilizável, salva as informações (command executor e ID) para ser reutilizado

        Returns
        ------
        `True` : bool
            Sucesso criando chrome
        `False` : bool
            Falha ao criar o chrome
        """

        # se o ID não existir...
        if not os.path.exists(self.id_path):
            logger.critical(f"Arquivo {self.id_path} não existe, criando um novo...")
            with open(self.id_path, "w", encoding="utf-8") as json_file:
                # cria um arquivo json com um dict vazio
                json.dump(dict(), json_file, indent=4)
            logger.critical(f"Arquivo '{self.id_path}' foi criado")

        # -----------------------------------
        # carrega o command executor e o session ID do arquivo id.json
        with open(self.id_path, "r") as json_file:
            try:
                chrome_ids = json.load(json_file)
            except json.decoder.JSONDecodeError:
                chrome_ids = dict()

        logger.debug(f"Arquivo de ID trouxe: '{chrome_ids}'")

        self.last_session_id = chrome_ids.get("session_id", None)
        self.last_command_executor = chrome_ids.get("command_executor", None)

        # -----------------------------------

        for tentativa_atual in range(self.attach_retries):
            logger.critical(f"Tentando criar driver {tentativa_atual+1}/{self.attach_retries}")
            attached = False

            # se tiver command_executor, session_id e um chrome existir, tenta usar uma sessão já aberta
            if (self.last_command_executor and self.last_session_id and self.any_chrome_process_exists):
                attached = self.__attach()

            # se não tiver um dos três pra dar attach, então precisa reabrir o chrome
            if not attached:
                self.end_all_driver_processes()  # encerra os chromes em execução
                self.__create_new_chrome()  # cria um novo chrome com Popen
                attached = self.__attach()  # tenta utilizar o chrome criado

            if attached:
                # se conseguiu dar attach em qualquer momento, sai do loop, se não continua
                break
        else:
            logger.critical(f"Não consegui criar um driver após '{self.attach_retries}' tentativas")
            return False

        # -----------------------------------
        # salva as informações da nova sessão do chromedriver
        chrome_configs = {
            "command_executor": self.command_executor._url,
            "session_id": self.session_id,
        }

        with open(self.id_path, "w", encoding="utf-8") as json_file:
            json.dump(chrome_configs, json_file, indent=4)

        logger.debug(f"Arquivo ID foi atualizado com: {chrome_configs}")

        return True

    def __create_new_chrome(self):
        """Cria um novo chrome, porém o método é um pouco diferente do convencional\n

        O chromedriver.exe é executado como uma nova janela (geralmente)\n
        Assim após o Python terminar a execução, o processo do chromedriver continuará ativo
        para podermos reutilizarmos posteriormente
        """
        logger.debug("Executando um novo Chrome")

        if self.new_console:
            flag = CREATE_NEW_CONSOLE
        else:
            flag = CREATE_NO_WINDOW

        full_launch = f'"{self.driver_path}" --port={self.port}'
        logger.debug(f"Executando chrome: '{full_launch}'")
        self.chrome_process = Popen(full_launch, creationflags=flag)

        self.last_command_executor = f"http://127.0.0.1:{self.port}"
        self.last_session_id = (
            None  # reseta o session ID, assim não utiliza o do arquivo
        )
        return

    def __attach(self):
        logger.debug("Tentando attachear a um Chrome já existente...")

        # pra dar o attach, tem que existir algum chrome aberto
        # então procura um processo do chrome
        if not self.any_chrome_process_exists:
            logger.debug("Nenhum processo com nome de chrome encontrado, não vou tentar dar attach...")
            return False

        # --------------------  --------------------

        if self.options is not None:
            # se a primeira vez que foi instanciado foi com user-data-dir,
            # na reutilização será necessário retirar esse parâmetro
            for opt in self.options.to_capabilities()["goog:chromeOptions"]["args"]:
                if "user-data-dir" in opt:
                    data_dir = opt.split("=")[1]

                    # se a pasta data-dir estiver em uso, remove o argumento para não dar erro
                    if os.path.exists(data_dir):  # se a pasta existe
                        # tenta renomear pro mesmo nome, valida se está em uso
                        # se der erro, significa que ela está em uso e o parâmetro precisa ser removido
                        try:
                            os.rename(data_dir, data_dir)
                        except OSError:
                            # retira a opção de data-dir do self.options
                            self.options.to_capabilities()["goog:chromeOptions"]["args"].remove(opt)
                    break

        # --------------------  --------------------

        try:
            # antigo pra referencia:
            # driver = webdriver.Remote(self.last_command_executor, options=self.options)

            # o super desta classe é o Remote
            super().__init__(command_executor=self.last_command_executor, options=self.options)

        except SessionNotCreatedException as e:
            logger.exception(f"Geralmente o chrome está desatualizado:")
            # tira a versão que o chrome está para baixar o chromedriver dela
            version = re.findall("version is (.*) with binary", str(e))
            if version:
                self.driver_path = ChromeDriverManager(path=self.driver_path).install()
            return False
        except (MaxRetryError, WebDriverException):
            logger.exception(f"Exceção ao conectar no chrome usando Remote:")
            self.end_all_driver_processes()
            return False

        # --------------------  --------------------

        # se não tem session ID então o chrome que foi aberto é novo
        # se já tinha session ID precisa fechar o driver (ele sempre abre um novo, mesmo usando remote, ele que será fechado) e mudar o ID
        if self.last_session_id is not None:
            self.close()
            self.session_id = self.last_session_id

        # --------------------  --------------------
        # validações pra verificar se o driver que foi aberto está responsivo
        try:
            for handle in self.window_handles:
                self.switch_to.window(handle)
                break
        except InvalidSessionIdException:
            logger.critical(f"Session ID '{self.last_session_id}' gravada não serve para o chrome atual")
            return False
        except WebDriverException:
            logger.critical("Geralmente chrome not reachable")
            return False

        # --------------------  --------------------
        # muda o implicity_wait do driver
        self.implicitly_wait(self.implicity_wait)
        # se ele falhar no Remote ou no swith_to, retorna False
        # pois não conseguiu de fato assumir o controle de um chrome
        # return False

        # --------------------  --------------------
        logger.critical("Attach com sucesso!")
        return True


class CreateChrome(CustomWebDriver, Chrome):

    def __init__(
        self,
        driver_path: Union[Path,
                           str],
        implicity_wait: int = 0,
        port: int = 64900,
        options: Optional[ChromeOptions] = None,
        **kwargs,
    ) -> None:

        if isinstance(driver_path, Path):
            driver_path = str(driver_path)
            cache = DriverCacheManager(driver_path)
        self.driver_path = ChromeDriverManager(cache_manager=cache).install()

        self.implicity_wait = implicity_wait
        self.port = port
        self.options = options
        return

    def begin(self):

        serv = ChromeService(executable_path=self.driver_path, port=self.port)

        super().__init__(service=serv, options=self.options)

        self.implicitly_wait(self.implicity_wait)

        return True

# --------------------------------------------------

def create_chrome():
    """This is an example function that you can copy and customize for each project"""
    BASE_DIR = Path(__file__).parent.resolve()
    logging.getLogger("selenium").setLevel(logging.WARNING)
    logging.getLogger("urllib3").setLevel(logging.WARNING)

    chrome_configs = {
        "driver_path": BASE_DIR / "/chromedriver",
        "id_path": BASE_DIR / "chromedriver/id.json",
        "port": 65000,
        "implicity_wait": 0,
        "attach_retries": 2,
        "new_console": True,
    }

    chrome_options = {
        "arguments": [
            "log-level=3",
            "no-first-run",
            # "incognito",
            "no-default-browser-check",
            "disable-infobars",
            "disable-blink-features",
            "disable-blink-features=AutomationControlled",
            rf"user-data-dir={BASE_DIR}\chromedriver\data-dir",
        ],
        "experimental": {
            "prefs": {
                "download.prompt_for_download": False,
                "download.directory_upgrade": True,
                "plugins.always_open_pdf_externally": True,
                "download.default_directory": BASE_DIR / "chromedriver/downloads",
            },
            "excludeSwitches": ["enable-automation", "ignore-certificate-errors"],
            "useAutomationExtension": False,
        },
        "extensions": [  
            # r"path_to\extensions" # --> a folder will use all .crx of folder
            # r"path_to\extensions\uBlock-Origin.crx", # --> only a single extension
        ],
    } # yapf: disable

    options = ChromeOptions()
    
    # --------------------------------------------------
    # add arguments
    for arg in chrome_options["arguments"]:
        options.add_argument(arg)

    # --------------------------------------------------
    # add experimental options
    for k, v in chrome_options["experimental"].items():
        options.add_experimental_option(k, v)

    # --------------------------------------------------
    # if extension isn't a list or a tuple, fixes to a list
    if chrome_options["extensions"] and not isinstance(chrome_options["extensions"], (list, tuple)):
        chrome_options["extensions"] = list(chrome_options["extensions"])

    # add extensions
    all_extensions = list()
    for ext in chrome_options["extensions"]:
        ext = Path(ext)
        if ext.is_dir():
            for e in ext.glob("*.*"):
                all_extensions.append(str(e))
        else:
            all_extensions.append(str(ext))

    for ext in all_extensions:
        options.add_extension(ext)

    # --------------------------------------------------

    # CustomChrome.end_all_chrome_processes()
    driver = CreateChrome(**chrome_configs, options=options) # creates a chrome that closes after program ends
    # driver = ReusableChrome(**chrome_configs, options=options) # creates a chrome session that can be reused over runs
    if not driver.begin():
        logger.critical(f"Something went wrong creating a chrome instance. Check logs for details.")
        return False

    # --------------------------------------------------

    driver.rotate_user_agent()

    driver.set_navigator_to_undefined()

    # driver.refresh()

    # corrects the exception: "unknown command: session/_____/se/file"
    driver._is_remote = False

    return driver
